metagraph/src/graph/alignment/aligner_helper.cpp

```diff
@@ -402,41 +403,58 @@ void Alignment<NodeType>::reverse_complement(const DeBruijnGraph &graph,
         // if the alignment starts from a source k-mer, then traverse forwards
         // until a non-dummy k-mer is hit and check if its reverse complement exists
         const auto *dbg_succ = dynamic_cast<const DBGSuccinct*>(&graph);
+        const auto *canonical = dynamic_cast<const CanonicalDBG*>(&graph);
+        if (!dbg_succ && canonical)
+            dbg_succ = dynamic_cast<const DBGSuccinct*>(&canonical->get_graph());
+
         if (dbg_succ && rev_seq[0] == boss::BOSS::kSentinel) {
-            const auto &boss = dbg_succ->get_boss();
-            assert(rev_seq.length() == graph.get_k());
-            auto edge = dbg_succ->kmer_to_boss_index(nodes_.back());
-            auto edge_label = boss.get_W(edge) % boss.alph_size;
-
-            // TODO: for efficiency, the last outgoing edge is taken at each step.
-            // This is valid for canonical graphs since the reverse complement of
-            // the found non-dummy k-mer is guaranteed to exist, but this node may
-            // not exist in a non-canonical graph.
-            for (size_t i = 0; i < offset_; ++i) {
-                edge = boss.fwd(edge, edge_label);
-                edge_label = boss.get_W(edge) % boss.alph_size;
-                if (edge_label == boss::BOSS::kSentinelCode) {
-                    // reverse complement not found
+            assert(nodes_.size() == 1);
+            nodes_[0] = DeBruijnGraph::npos;
+            // TODO: find a better way to pick one node
+            dbg_succ->call_nodes_with_prefix_matching_longest_prefix(
+                sequence_,
+                [&](NodeType prefix_node, size_t) { nodes_[0] = prefix_node; },
+                sequence_.size(),
+                [&]() -> bool { return nodes_[0]; }
+            );
+
+            if (nodes_[0]) {
+                rev_seq = dbg_succ->get_node_sequence(nodes_[0]);
+                std::vector<NodeType> rev_nodes = nodes_;
+                reverse_complement_seq_path(graph, rev_seq, rev_nodes);
+
+                assert(std::find(rev_nodes.begin(), rev_nodes.end(),
+                                 DeBruijnGraph::npos) == rev_nodes.end());
+
+                assert(rev_seq.size() > offset_);
+                sequence_.assign(rev_seq.begin() + offset_, rev_seq.end());
+                nodes_.assign(rev_nodes.begin(), rev_nodes.end());
```

Comment:
nodes_ always has size 1, right? 

Why do you treat it as a vector with multiple elements? (e.g., std::vector<NodeType> rev_nodes = nodes_;)